### PR TITLE
Open VSX Registry integration

### PR DESCRIPTION
**Pull request by jbicker**
#### What it does
Widget where one can search and install VSCode Extensions from the Open VSX Registry.

Screenshot 2020-03-11 at 12 36 59

TODO:
- [x] keyboard navigation does not work (rewrite lists with trees)
- [x] installed extensions metadata should not be fetched from the registry, but should use local
- [x] vscode extension URI resolver should be installed, see https://github.com/eclipse-theia/theia/pull/6655#issuecomment-593582431
- [x] extensions should be installed under `${configDir}/extensions` by default
- [x] extensions should be loaded from `${configDir}/extensions` by default
- [x] focus is not properly given on the view activation
- [x] support built-in extensions
- [x] update view container layout based on query, preserve layout state for different modes
- [x] progress reporting in view container for search, install, uninstall
- [x] theming
- [x] preserve state properly, i.e. search query should be preserved only, everything else should be computed from it again
- [x] check TODOs in code

Out of scope:
- finding a version compatible extension
- detecting outdated extensions and upgrading them
- supporting workspace extensions (only built-in and global (user))
- supporting extension enablement
- use webview to render remote content like README

#### How to test

- Try to install/uninstall extensions. Make sure that they are preserved (or not after uninstall) on restart.
- Look at the extension editor.
- Test that the extension editor is opened for the vscode extension URIs.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/vscode-extensions/src/browser/vscode-extensions-service.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import { VSCodeExtensionUri, VSCodeExtensionDetailOpenerOptions } from './view/detail/vscode-extension-open-handler';
+import { PluginServer } from '@theia/plugin-ext';
+import { HostedPluginSupport } from '@theia/plugin-ext/lib/hosted/browser/hosted-plugin';
+
+export type ExtensionKeywords = string[];
+export const ExtensionKeywords = Symbol('ExtensionKeyword');
+
+// for now to test the extension one has to start the Registry Server via https://gitpod.io/#https://github.com/theia-ide/extension-registry
+// and copy then the workspace url plus '/api' to here
+export const API_URL = 'https://8080-c5484d3a-4b1a-4f66-81aa-c5fbc0d8ee0e.ws-eu01.gitpod-staging.com/api';
```
Comment (svenefftinge) [inline packages/vscode-extensions/src/browser/vscode-extensions-service.ts]: We should host an instance somehow until we have the one under eclipse.org running.
Also this should become a preference.

**Comment by svenefftinge** [comment]
The items should have an 'install' button when not installed, and an 'uninstall' button when installed.

**Comment by svenefftinge** [comment]
The progress during installation should be shown with blue moving line above the search box, just as it does in other similar views. The button should turn to 'installing' during that.

**Comment by svenefftinge** [comment]
I cannot install any extension. The button spins forever.

**Comment by svenefftinge** [comment]
The search doesn't seem to work. If I search for 'Docker' I see many unrelated extensions.

**Comment by svenefftinge** [comment]
Screenshot 2019-11-28 at 15 14 52

**Comment by jbicker** [comment]
Regarding search: https://github.com/theia-ide/extension-registry/issues/3

File: packages/vscode-extensions/src/browser/style/extension-detail.css [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+.vscode-extension-detail {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/style/extension-detail.css]: please check that code in PR is aligned with our coding conventions

i.e here naming is bogus: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#css-prefix-global-classes

File: packages/vscode-extensions/package.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    "@types/sanitize-html": "^1.13.31",
+    "@types/showdown": "^1.7.1",
+    "sanitize-html": "^1.14.1",
+    "showdown": "^1.9.1"
+  },
+  "publishConfig": {
+    "access": "public"
+  },
+  "keywords": [
+    "theia-vscode-extensions"
```
Comment (akosyakov) [inline packages/vscode-extensions/package.json]: It should be the same as in other packages

File: packages/vscode-extensions/src/browser/style/extensions.svg [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<!--Copyright (c) Microsoft Corporation. All rights reserved.-->
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/style/extensions.svg]: Please file a cq for all copied code

File: packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            return uri.fragment;
+        }
+        throw new Error('The given uri is not an vscode extension URI, uri: ' + uri);
+    }
+}
+
+export type VSCodeExtensionDetailOpenerOptions = WidgetOpenerOptions & VSCodeExtensionDetailWidgetOptions;
+
+@injectable()
+export class VSCodeExtensionOpenHandler extends WidgetOpenHandler<VSCodeExtensionDetailWidget> {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts]: We should make sure that extension uris conform to VS Code, that links from webviews work properly

File: packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        try {
+            VSCodeExtensionUri.toExtensionName(uri);
+            return 500;
+        } catch {
+            return 0;
+        }
+    }
+
+    protected createWidgetOptions(uri: URI, options: VSCodeExtensionDetailOpenerOptions): VSCodeExtensionDetailWidgetOptions {
+        return options;
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts]: It looks wrong, it should be options which are uniquely identify widgets

File: packages/vscode-extensions/src/browser/view/vscx-install-button-component.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                        this.onUninstallButtonClicked();
+                    } else {
+                        this.onInstallButtonClicked();
+                    }
+                    event.stopPropagation();
+                }
+            }}
+        >{content}</div>);
+
+        // if (extension.outdated) {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/view/vscx-install-button-component.tsx]: Please remove commented out code

File: packages/vscode-extensions/src/browser/vscode-extensions-api.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
+/********************************************************************************
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/vscode-extensions-api.ts]: File name and api name is confusing, I believe I does not have anything to do with vscode extension apis, but with registry api

File: packages/vscode-extensions/src/browser/vscode-extensions-api.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import { VSCodeExtensionPart, VSCodeExtensionFull, VSCodeExtensionReviewList } from './vscode-extensions-types';
+
+export interface ExtensionRegistryAPIRequest<T> {
+    endpoint: string,
+    operation: (response: Response) => Promise<T>
+    method: 'GET'
+}
+
+@injectable()
+export class VSCodeExtensionsAPI {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/vscode-extensions-api.ts]: VSXRegistryService

File: packages/vscode-extensions/src/browser/vscode-extensions-contribution.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    export const CLEAR_ALL: Command = {
+        id: 'vscode-extensions.clear-all',
+        category: 'Extensions',
+        label: 'Clear Search Results',
+        iconClass: 'clear-all'
+    };
+}
+
+@injectable()
+export class VSCodeExtensionsContribution extends AbstractViewContribution<VSCodeExtensionsWidget> {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/vscode-extensions-contribution.ts]: VSXRegistryContribution, please make sure in the rest of PR that it is about interactions with the registry, not implementing vscode eztensions

File: packages/vscode-extensions/src/browser/vscode-extensions-service.ts (hunk 1 above)
Comment (akosyakov) [reply packages/vscode-extensions/src/browser/vscode-extensions-service.ts]: Not sure about preference, should a user be able to change it? I think it should be configurable for the prpduct

File: packages/vscode-extensions/src/browser/vscode-extensions-service.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    protected async compileDocumentation(extension: VSCodeExtensionFull): Promise<string> {
+        if (extension.readmeUrl) {
+            const markdownConverter = new showdown.Converter({
+                noHeaderId: true,
+                strikethrough: true,
+                headerLevelStart: 2
+            });
+            const readme = await this.api.getExtensionReadMe(extension.readmeUrl);
+            const readmeHtml = markdownConverter.makeHtml(readme);
+            return sanitize(readmeHtml, {
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/vscode-extensions-service.ts]: I’ve noticed that instead VS Code shows details in webview. cc @svenefftinge

File: packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            VSCodeExtensionUri.toExtensionName(uri);
+            return 500;
+        } catch {
+            return 0;
+        }
+    }
+
+    protected createWidgetOptions(uri: URI, options: VSCodeExtensionDetailOpenerOptions): VSCodeExtensionDetailWidgetOptions {
+        return {
+            url: options.url
```
Comment (akosyakov) [inline packages/vscode-extensions/src/browser/view/detail/vscode-extension-open-handler.ts]: `uri: uri.trimFragment().toString()`

many different urls can be used to fetch the same uri, e.g. think different registry, different versions for `vscode:extension/ms-vscode.Go`

**Comment by akosyakov** [comment]
- [ ] respect extensions vscode colors, see https://code.visualstudio.com/api/references/theme-color#extensions

**Comment by akosyakov** [comment]
Please clean up the history: https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#checklist-commit-history There are many commits which fixes introduced by this PR. If it is hard then squash it.

File: packages/vsx-registry/src/browser/vsx-registry-contribution.ts [hunk 13]
```diff
@@ -58,6 +60,29 @@ export class VSXRegistryContribution extends AbstractViewContribution<VSXRegistr
         });
     }
 
+    async registerColors(colors: ColorRegistry): Promise<void> {
+        colors.register(
```
Comment (akosyakov) [inline packages/vsx-registry/src/browser/vsx-registry-contribution.ts]: please like corresponding section of VS Code colors references

File: packages/vsx-registry/src/browser/vsx-registry-service.ts [hunk 14]
```diff
@@ -32,7 +32,8 @@ export const ExtensionKeywords = Symbol('ExtensionKeyword');
 
 @injectable()
 export class VSXRegistryService {
-    protected apiUrl: string;
+    apiUrl: string;
```
Comment (akosyakov) [inline packages/vsx-registry/src/browser/vsx-registry-service.ts]: `string | undefined`? clients can change it?

File: packages/vsx-registry/src/browser/vsx-registry-contribution.ts (hunk 13 above)
Comment (jbicker) [reply packages/vsx-registry/src/browser/vsx-registry-contribution.ts]: What do you mean exactly?

Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-registry-contribution.ts]: sorry, I meant "link"

**Comment by akosyakov** [comment]
I have an issue that vsx-registry extension should provide a resolver for `vscode:extension/publisher.name` URIs, but we already resolve them in plugin-ext-vscode extension from VS Code marketplace: https://github.com/eclipse-theia/theia/blob/4c04d02910f4ba86938c54b8094c2dc257148bca/packages/plugin-ext-vscode/src/node/plugin-vscode-resolver.ts#L30

How should it be handled? cc @marcdumais-work @svenefftinge @benoitf @tsmaeder 

Do we still allow Theia to access VS Code marketplace? Or we remove it to avoid violating VS Code Markteplace TOS by end products and default to our registry?

**Comment by tsmaeder** [comment]
@akosyakov in principle, I think we should remove the resolution of plugins via VS Code marketplace: it's simply not the right thing to do. That said, I believe in Che we have a couple of cases where some plugins rely on resolution via the marketplace (even though they shouldn't). Maybe @benoitf knows off the top of his head what the state is there. Otherwise, I propose we do an investigation to clear that up. In any case, I think Che would appreciate it if we could coordinate on the removal of marketplace resolution.
On a more general note, I think providing a concrete marketplace should be on the level of products, not the Theia framework. So I would move provisioning from the theia marketplace to a separate package and not have a default marketplace at all. 
In general, we would benefit from separating the installation of plugins from running the IDE more: for example in Che, we have a separate plugin broker that should be provisioning plugins inside their proper containers, not theia itself. Installation should be push, not pull as far as the IDE is concerned, IMO.

**Comment by akosyakov** [comment]
> On a more general note, I think providing a concrete marketplace should be on the level of products, not the Theia framework. So I would move provisioning from the theia marketplace to a separate package and not have a default marketplace at all.

@tsmaeder yes, it was also thought:
- remove resolution from `plugin-ext-vscode` against VS Code Marketplace (no resolution by default)
- `vsx-registry` extension has own resolution against VSX Open Registry hosted by Eclipse Foundation, it allows to reconfigure to use another registry URI on product and user levels

So all products which don't use `vsx-registry` extension, still can provide own resolution against custom marketplaces.

> In any case, I think Che would appreciate it if we could coordinate on the removal of marketplace resolution.

That's I'm asking. I've added it to the dev meeting.

**Comment by akosyakov** [comment]
It was discussed during the dev meeting to proceed how it is outlined in https://github.com/eclipse-theia/theia/pull/6655#issuecomment-593853221.

**Comment by akosyakov** [comment]
I wonder how we should distinguish built-in extensions from one which can be uninstalled. cc @marcdumais-work @svenefftinge @tsmaeder @benoitf 

Should we base it on env variables? i.e. if extensions is contributed using `THEIA_DEFAULT_PLUGINS`  then there are built-in and it is not allowed to uninstall them, otherwise not?

**Comment by marcdumais-work** [comment]
> Should we base it on env variables? i.e. if extensions is contributed using `THEIA_DEFAULT_PLUGINS` then there are built-in and it is not allowed to uninstall them, otherwise not?

That sounds reasonable. The need to have extensions that are not allowed to be deleted or updated may not be limited to built-ins - a product designer may not wish the users to mess-with other VS Code extensions bundled with the product. Then the same mechanism could be used. (e.g.  older versions of Java by Red Hat work in Theia, but later versions of that extension do not, so we deliver a working version with product that should stay as-is)

> I wonder how we should distinguish built-in extensions from one which can be uninstalled

So maybe distinguish between extensions that are allowed to be removed or updated, and those that are not allowed, rather than built-ins vs others? Whatever extensions refered by env var `THEIA_DEFAULT_PLUGINS` are considered "read-only", and those refered by `THEIA_PLUGINS` are allowed to be un-installed/updated?

**Comment by akosyakov** [comment]
> That sounds reasonable. The need to have extensions that are not allowed to be deleted or updated may not be limited to built-ins - a product designer may not wish the users to mess-with other VS Code extensions bundled with the product. Then the same mechanism could be used. (e.g. older versions of Java by Red Hat work in Theia, but later versions of that extension do not, so we deliver a working version with product that should stay as-is)

It would be nice to group built-in extensions in the separate view from installed and search. I think if something should be readonly then it should be built-in, whether built-in is installed from the registry or bundled should not be matter.

> So maybe distinguish between extensions that are allowed to be removed or updated, and those that are not allowed, rather than built-ins vs others? Whatever extensions refered by env var THEIA_DEFAULT_PLUGINS are considered "read-only", and those refered by THEIA_PLUGINS are allowed to be un-installed/updated?

Yes, i was thinking like that as well, but then it is not easy to map and different product can pull built-in extensions from different places like the registry or own mechanism, not necessary environment variables. Now I think maybe we just list them in the declarative way in package.json? So it is decoupled from how they delivered.

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 15]
```diff
@@ -675,6 +678,8 @@ export interface PluginDeployerHandler {
     deployFrontendPlugins(frontendPlugins: PluginDeployerEntry[]): Promise<void>;
     deployBackendPlugins(backendPlugins: PluginDeployerEntry[]): Promise<void>;
 
+    removeDeployedPluginId(pluginId: string): void;
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/plugin-protocol.ts]: `undeployPlugin`

File: packages/vsx-registry/.gitignore [hunk 16]
```diff
@@ -0,0 +1 @@
+*-test-temp
```
Comment (akosyakov) [inline packages/vsx-registry/.gitignore]: remove

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+        }
+        return <div className='theia-vsx-extension'>
+            <div className='theia-vsx-extension-icon' style={iconStyle} />
+            <div className='theia-vsx-extension-content'>
+                <div className='title'>
+                    <div className='noWrapInfo'>
+                        <span className='name'>{displayName}</span> <span className='version'>{version}</span>
+                    </div>
+                    <div className='stat'>
+                        {downloadCount && <span className='fa fa-download download-count'>{downloadCount}</span>}
```
Comment (akosyakov) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: Number(downloadCount).toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,')

File: packages/vsx-registry/README.md [hunk 18]
```diff
@@ -0,0 +1,4 @@
+# Theia - VSCode Extensions
```
Comment (akosyakov) [inline packages/vsx-registry/README.md]: align with @vince-fugnitto template

document environment variable to configure vsx registry URL.

**Comment by marcdumais-work** [comment]
Hi @akosyakov 

> It would be nice to group built-in extensions in the separate view from installed and search. I think if something should be readonly then it should be built-in, whether built-in is installed from the registry or bundled should not be matter.

> Yes, i was thinking like that as well, but then it is not easy to map and different product can pull built-in extensions from different places like the registry or own mechanism, not necessary environment variables. Now I think maybe we just have a hook to indicate which plugin is built-in? It could fallback to some declarative configuration for electron products. So it is decoupled from how they delivered.

Sounds good to me. I was mostly thinking that we would support bundling built-in extensions in products, but if we can as well make it so that these extensions can be installed at run-time, that's even better.

**Comment by akosyakov** [comment]
@marcdumais-work I have a following idea so far:
- there will be a new enumeration `PluginType` which either has System or User value, depending who triggered installation an embedding product or a user
- based on it `PluginVsCodeFileHandler` and `PluginTheiaFileHandler` will decide where to unarchive:
  - if a type is `System` then as before
  - if a type is `User` then under user home in extensions and plugins folder correspondingly. With `User` type there will be as well a check whether we already unarchived it before.
- On start we will add extensions and plugins folder as local-dir for deployment always with `User` type, so plugins/extensions installed by a user are restored . For all other plugins coming from env variables like `THEIA_DEFAULT_PLUGIN` and `THEIA_PLUGINS` logic will remain the same by deploying with `System` type.
- `PluginDeployer.deploy` method will accept a type and UI will pass `User` as a type

In short, by default:
- if a product does not allow plugin installations from Theia UI everything will be work the same way as now
- if a product though allows to install then plugins will be persisted in the user config location and loaded from there on next start. Products can change it by overriding `PluginDeployer.deploy` to apply `System` type always then plugins installed by a user will be lost after restart as on master.

@benoitf sounds good?

**Comment by benoitf** [comment]
@akosyakov yes

**Comment by akosyakov** [comment]
The current state:
Screenshot 2020-03-11 at 12 36 59

Hope to move it for a review today or tomorrow.

**Comment by akosyakov** [comment]
@benoitf, @azatsarynnyy I've extracted changes to the plugin system into separate commits. Could you have a look at them?

File: packages/plugin-ext-vscode/src/node/scanner-vscode.ts [hunk 19]
```diff
@@ -43,7 +43,10 @@ export class VsCodePluginScanner extends TheiaPluginScanner implements PluginSca
             },
             entryPoint: {
                 backend: plugin.main
-            }
+            },
+            iconUrl: plugin.icon && PluginPackage.toPluginUrl(plugin, plugin.icon),
```
Comment (akosyakov) [inline packages/plugin-ext-vscode/src/node/scanner-vscode.ts]: I've added it only for VS Code extensions, but I think it it should be useful for Theia plugins as well, i.e. if in che you have a detail widget for plugin it will be useful to fetch which data. @benoitf @azatsarynnyy Should I align Theia scanner as well?

**Comment by akosyakov** [comment]
@marcdumais-work @svenefftinge @vince-fugnitto I'm waiting for @spoenemann deployment of a never version of vsx-registry to align integration and after that it should be good for testing.

File: packages/plugin-ext-vscode/package.json [hunk 20]
```diff
@@ -10,6 +10,7 @@
     "@theia/plugin-ext": "^0.16.0",
     "@theia/workspace": "^0.16.0",
     "@types/request": "^2.0.3",
+    "filenamify": "^4.1.0",
```
Comment (vince-fugnitto) [inline packages/plugin-ext-vscode/package.json]: If not already done so, we will need to verify using the [process](https://github.com/eclipse-theia/theia/wiki/Registering-CQs#case-new-or-updated-npm-production-dependency) if the new dependency is license-certified.

Comment (akosyakov) [reply packages/plugin-ext-vscode/package.json]: it is MIT as well as its dependencies

**Comment by akosyakov** [comment]
> Also, moving the view sometimes hides content when I believe the goal is to keep the item as completely visible by minimum:

It is not easy, we can hardcode a value equal to high of a record, not sure whether it is desirable for empty view.

**Comment by vince-fugnitto** [comment]
> > Also, moving the view sometimes hides content when I believe the goal is to keep the item as completely visible by minimum:
> 
> It is not easy, we can hardcode a value equal to high of a record, not sure whether it is desirable for empty view.

I think it's very minor, we can improve it at a later date :)

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+@injectable()
+export class VSXExtensionsWidgetOptions extends VSXExtensionsSourceOptions {
+}
+
+@injectable()
+export class VSXExtensionsWidget extends SourceTreeWidget {
+
+    static ID = 'vsx-extensions';
+    static INSTALLED_ID = VSXExtensionsWidget.ID + ':' + VSXExtensionsSourceOptions.INSTALLED;
```
Comment (akosyakov) [inline packages/vsx-registry/src/browser/vsx-extensions-widget.ts]: used?

**Comment by akosyakov** [comment]
Latest version of registry was deployed, the PR should be aligned with it. I will have a look at an icon issue tomorrow.

@vince-fugnitto I've set a min height for extension rows.

File: packages/vsx-registry/src/common/vsx-registry-api.ts [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+                query += 'size=' + param.size;
+            }
+            if (param.offset) {
+                query += 'offset=' + param.offset;
+            }
+            if (query) {
+                searchUri = searchUri.withQuery(query);
+            }
+        }
+        return this.fetchJson<VSXSearchResult>(searchUri.toString());
```
Comment (spoenemann) [inline packages/vsx-registry/src/common/vsx-registry-api.ts]: For some reason, the whole query is URI-encoded. If I type `foo`, the request to the registry is `https://open-vsx.org/api/-/search?query%3Dfoo`. The `=` character should not be encoded!

File: packages/vsx-registry/src/common/vsx-registry-types.ts [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * Should be aligned with https://github.com/eclipse/openvsx/blob/master/server/src/main/java/org/eclipse/openvsx/json/ExtensionJson.java
+ */
+export interface VSXExtensionRaw {
+    readonly error?: string;
+    readonly namespaceUrl: string;
+    readonly reviewsUrl: string;
+    readonly name: string;
+    readonly namespace: string;
```
Comment (spoenemann) [inline packages/vsx-registry/src/common/vsx-registry-types.ts]: Add `publishedBy` and `namespaceAccess` properties:
https://github.com/eclipse/openvsx/blob/74dd60eebc6855411b91f67ff809c15d9ad68a7a/server/src/main/java/org/eclipse/openvsx/json/ExtensionJson.java#L43-L46

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 24]
```diff
@@ -0,0 +1,441 @@
...
+                <div className='icon-container' style={iconStyle} />
+                <div className='details'>
+                    <div className='title'>
+                        <span title='Extension name' className='name' onClick={this.openExtension}>{displayName}</span>
+                        <span title='Extension identifier' className='identifier'>{id}</span>
+                        {preview && <span className='preview'>Preview</span>}
+                        {builtin && <span className='builtin'>Built-in</span>}
+                    </div>
+                    <div className='subtitle'>
+                        <span title='Publisher name' className='publisher' onClick={this.searchPublisher}>{publisher}</span>
```
Comment (spoenemann) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: Please align with open-vsx.org and display the `publishedBy` info here instead of the publisher, along with an icon indicating the `namespaceAccess` info (`public` or `restricted`).

https://github.com/eclipse/openvsx/blob/74dd60eebc6855411b91f67ff809c15d9ad68a7a/webui/src/pages/extension-detail/extension-detail.tsx#L200-L242

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 25]
```diff
@@ -0,0 +1,441 @@
...
+                        {downloadCount && <span className='fa fa-download download-count' onClick={this.openExtension}>{downloadFormatter.format(downloadCount)}</span>}
+                        {averageRating && <span className='average-rating' onClick={this.openAverageRating}>{this.renderStars()}</span>}
```
Comment (spoenemann) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: The separator between rating and repository is a bit higher than the other separators.

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 26]
```diff
@@ -0,0 +1,441 @@
...
+                        {repository && <span className='repository' onClick={this.openRepository}>Repository</span>}
```
Comment (spoenemann) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: Maybe display the repository type, e.g. "GitHub", instead of "Repository"?

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 27]
```diff
@@ -0,0 +1,441 @@
...
+                        {license && <span className='license' onClick={this.openLicense}>License</span>}
```
Comment (spoenemann) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: Display the license type (`license` property) instead of "License".

https://github.com/eclipse/openvsx/blob/74dd60eebc6855411b91f67ff809c15d9ad68a7a/webui/src/pages/extension-detail/extension-detail.tsx#L244-L254

File: packages/vsx-registry/src/browser/vsx-extension.tsx [hunk 28]
```diff
@@ -0,0 +1,441 @@
...
+        const iconStyle: React.CSSProperties = {};
+        if (iconUrl) {
+            iconStyle.backgroundImage = `url('${iconUrl}')`;
+        }
+        return <React.Fragment>
+            <div className='header'>
+                <div className='icon-container' style={iconStyle} />
+                <div className='details'>
+                    <div className='title'>
+                        <span title='Extension name' className='name' onClick={this.openExtension}>{displayName}</span>
```
Comment (spoenemann) [inline packages/vsx-registry/src/browser/vsx-extension.tsx]: A not-so-nice side effect of using `onClick` handlers instead of `<a>` elements is that the user cannot see where the link is going. I find it a bit confusing in this case because there are so many `onClick` links and they all lead to different targets.

File: packages/vsx-registry/src/browser/vsx-extension.tsx (hunk 26 above)
Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-extension.tsx]: It's aligned with VS Code.

File: packages/vsx-registry/src/browser/vsx-extension.tsx (hunk 24 above)
Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-extension.tsx]: I actually find open VSX UI confusing, publisher should be a filed from package.json, but it shows the person who published and never show `namespace` field at all.

I cannot do it here because I have to care about locally installed extensions as well and for them it is `publisher` field from package.json.

Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-extension.tsx]: I will render an icon before namespace to indicate access (globe for public, shield for restricted), title will explain and on click the actual publisher will be opened. For built-in extension there won’t be such icon.

File: packages/vsx-registry/src/browser/vsx-extension.tsx (hunk 28 above)
Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-extension.tsx]: There are different css classes are applied by default to elements in theia to how render them active, focused or visited. It would require undoing all of them.

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts (hunk 21 above)
Comment (akosyakov) [reply packages/vsx-registry/src/browser/vsx-extensions-widget.ts]: turned out to be used

**Comment by akosyakov** [comment]
I've rebased and addressed all comments from @vince-fugnitto and @spoenemann.

The current state:
Screenshot 2020-03-13 at 17 22 50

@marcdumais-work @svenefftinge please have a look

**Comment by akosyakov** [comment]
> the search does not work properly (ex: search for 'a' yields not results but 'alt' does)

@vince-fugnitto I've filed an issue for the registry: https://github.com/eclipse/openvsx/issues/34 cc @spoenemann 

I'm not sure about other issues, they seem to be minor. I would better invest time into integration tests for typescript againt new Monaco. Someone else can improve styling after this PR is merged.

**Comment by akosyakov** [comment]
I'm going to merge it tomorrow if no objection. Theia 1.0 is already next week.

**Comment by vince-fugnitto** [comment]
> > the search does not work properly (ex: search for 'a' yields not results but 'alt' does)
> 
> @vince-fugnitto I've filed an issue for the registry: [eclipse/openvsx#34](https://github.com/eclipse/openvsx/issues/34) cc @spoenemann

👍 

> I'm not sure about other issues, they seem to be minor. I would better invest time into integration tests for typescript againt new Monaco. Someone else can improve styling after this PR is merged.

I'm fine with that, the improvements can be made iteratively at a later point. 
I'll file issues for them once the pull-request is merged.

**Comment by vince-fugnitto** [comment]
With the **VSX** (Extension) view now merged, is there still a reason to have the **Plugins** view?
